File: 2022/05 - Supply Stacks/solution.py
#!/usr/bin/python3
# --- Day 5: Supply Stacks ---

import re

with open("input.txt") as input:
    data = input.read().strip()

def initializeStacks(stackDescription):
    return list(map(
        lambda s: list(filter(lambda item: item.isalpha(), s)),
        list(zip(*reversed(stackDescription.splitlines())))[1::4]
    ))

def moveCargo(craneType = 9000):
    stackDescription, instructions = data.split("\n\n", 1)
    stacks = initializeStacks(stackDescription)
    for instr in instructions.splitlines():
        count, fromStack, toStack = [int(n) for n in re.findall(r'\d+', instr)]
        cargo = stacks[fromStack-1][-count:]
        stacks[fromStack-1] = stacks[fromStack-1][:-count]
        if craneType == 9000:
            cargo.reverse()
        stacks[toStack-1].extend(cargo)
    return "".join([s[-1] for s in stacks])

silver = lambda: moveCargo(craneType = 9000)
gold = lambda: moveCargo(craneType = 9001)

if __name__ == "__main__":
    print("silver: %s" % silver())
    print("gold:   %s" % gold())

File: 2022/05 - Supply Stacks/solution.js
/* --- Day 5: Supply Stacks --- */

var silver = () => moveCargo(9000);
var gold = () => moveCargo(9001);

function initStacks(stackDesc) {
    let [stackIds, ...cargoList] = stackDesc.split(/\n/).reverse(),
        stacks = Array.from(new Array(stackIds.match(/\S+/g).length), () => []);
    cargoList.forEach(cargo =>
        stacks.forEach(
            (stack, i) =>
                cargo[i * 4 + 1] != ' ' && stack.push(cargo[i * 4 + 1])
        )
    );
    return stacks;
}

function moveCargo(craneType) {
    let [stackDesc, moves] = parseInput(/\n\n/),
        stacks = initStacks(stackDesc);
    moves
        .split(/\n/)
        .map(move => move.match(/\d+/g))
        .forEach(([cnt, from, to]) => {
            let temp = stacks[from - 1].splice(-cnt);
            if (craneType == 9000) temp.reverse();
            stacks[to - 1].push(...temp);
        });
    return stacks.map(s => s[s.length - 1]).join('');
}
